src/main/java/org/opensearch/ad/cluster/HashRing.java

```diff
@@ -243,17 +240,21 @@ private void buildCircles(Set<String> removedNodeIds, Set<String> addedNodeIds,
         try {
             DiscoveryNode localNode = clusterService.localNode();
             if (removedNodeIds != null && removedNodeIds.size() > 0) {
-                LOG.info("Remove nodes from AD version hash ring: {}", Arrays.toString(removedNodeIds.toArray(new String[0])));
+                LOG.info("Node removed: {}", Arrays.toString(removedNodeIds.toArray(new String[0])));
                 for (String nodeId : removedNodeIds) {
-                    removeNodeFromCircles(nodeId);
+                    ADNodeInfo nodeInfo = nodeAdVersions.remove(nodeId);
+                    if (nodeInfo != null && nodeInfo.isEligibleDataNode()) {
+                        removeNodeFromCircles(nodeId, nodeInfo.getAdVersion());
+                        LOG.info("Remove data node from AD version hash ring: {}", nodeId);
+                    }
```

Comment:
Can you confirm that only data nodes are added to the hash ring? And we store all versions regardless of data node or not?

Reply:
Yes, that's correct. Hash ring only stores eligible data nodes so when we dispatch realtime RCF model or historical task, we only choose one from those data nodes. But we store all nodes' AD version to make sure if the request from non-data-type node, we can recognize the its AD version and send back correct response.